Add ClientTransportFilter

joybestourous opened the pull request:
**Problem**
We're trying to add some stats on the number of connections opened, closed, and active on the client side for grpc-java. On the server side, ServerTransportFilter acts as a reasonable hook for this, but there doesn't seem to be anything comparable on the client side that I can find (if there is, please direct me to it!).

Closes #10647

**Solution**
Add a ClientTransportFilter. Similarly to ServerTransportFilter this will provide an observability hook and allows direct modification of the transport's attributes. 
To wire this in, this PR also adds a `filterTransport` method to `ManagedClientTransport.Listener` which handles modifying the transport's attributes. This is called just before `transportReady`

ejona86 commented on api/src/main/java/io/grpc/ForwardingChannelBuilder.java:
```diff
@@ -91,6 +91,12 @@ public T intercept(ClientInterceptor... interceptors) {
   }
 
+  @Override
+  public T addTransportFilter(ClientTransportFilter transportFilter) {
```
You can remove this one. ForwardingChannelBuilder extends ForwardingChannelBuilder2, so it'll be here already. We're set up to "never touch ForwardingChannelBuilder again."

ejona86 commented on core/src/main/java/io/grpc/internal/InternalSubchannel.java:
```diff
@@ -542,6 +547,9 @@ private class TransportListener implements ManagedClientTransport.Listener {
     public void transportReady() {
       channelLogger.log(ChannelLogLevel.INFO, "READY");
+      for (ClientTransportFilter filter : transportFilters) {
+        filter.transportReady();
```
If we were to pass the attributes, they can be gotten via `transport.getAttributes()`.

ejona86 commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -661,6 +664,7 @@ ClientStream newSubstream(
       channel = builder.binlog.wrapChannel(channel);
     }
     this.interceptorChannel = ClientInterceptors.intercept(channel, interceptors);
+    this.transportFilters = builder.transportFilters;
```
We need to make a copy so we know they aren't mutated. (Someone can add a filter after calling build(), then call build() a second time.)

ejona86 commented on api/src/main/java/io/grpc/ClientTransportFilter.java:
```diff
@@ -0,0 +1,55 @@
+ * for transport observability. Multiple filters may be registered to the client.
+ */
+@ExperimentalApi("https://gitub.com/grpc/grpc-java/issues/TODO")
+public abstract class ClientTransportFilter {
```
"Filter." The server-side piece is called a Filter because it receives the transport's attributes and can change them. They are eventually viewable via ServerCall.getAttributes(). This allows you to add some data in transportReady and see it again in transportTerminated. That looks like it may be annoying to plumb here, but I'd have to stare at it a bit longer to tell.

I don't necessarily think we really need that though. We added the filtering for a specific purpose (caching data per-connection), which then went nowhere. Almost certainly it seems the filters would want to know the IP addresses and such of a transport.

sergiitk commented on api/src/main/java/io/grpc/ClientTransportHook.java:
```diff
@@ -0,0 +1,48 @@
+ * to modify the channels or transport life-cycle event behavior, but they can be useful hooks
+ * for transport observability. Multiple filters may be registered to the client.
+ */
+@ExperimentalApi("https://gitub.com/grpc/grpc-java/issues/TODO")
```
Same as the other comment:

> This should be a new ticket with "experimental API"  label. Example: https://github.com/grpc/grpc-java/issues/10108.

I think it'd even make sense to use the same ticket to track both `ClientTransportHook` interface, and channel builder's `addTransportHook()`. 

@ejona86 - correct me if I'm wrong here

sergiitk commented on api/src/main/java/io/grpc/ClientTransportHook.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Listens on the client transport life-cycle events. These filters do not have the capability
+ * to modify the channels or transport life-cycle event behavior, but they can be useful hooks
+ * for transport observability. Multiple filters may be registered to the client.
```
```suggestion
 * for transport observability. Multiple filters may be registered to the client.
 *
 * @since 1.60.0
```

joybestourous commented:
👋 @ejona86 any update on this PR? is this still on track for v61 release?

ejona86 commented:
(FYI, the branch cut for 1.61 is a week earlier than normal, or maybe a bit more. So it is coming up soon. I'd be fine backporting this to the release, even though it is technically a feature, if we can do so at least a week before the release. If there's any bugs, seem they should only impact you and not destabilize others.)

joybestourous commented:
> Sorry I went silent on this. I had been hoping to sneak it into 1.60 but then issues with the release cropped up that ended up taking a lot of my time. (I actually didn't see your comment two days ago either...)
> 
> We discussed in our API review and the general verdict was "make it a filter like on server-side." Having it just be a hook caused a whole slew of naming/identity/purpose crises that are all solved by mirroring the existing server-side API. Before my time evaporated I was going to actually make the necessary plumbing change...
> 
> The listener could receive and return Attributes, just like is done on server-side. In OkHttpClientTransport, BinderTransport, InProcessTransport, CronetClientTransport look to be pretty easy to support (some places attributes would no longer be final; that's okay). NettyClientHandler could pass the Attributes through ClientTransportLifecycleManager.nofiyReady() and update the attributes on return.
> 
> But I think that would be inherently racy, in a way not present on server-side. `transportReady()` is when the client can start sending RPCs on the transport, so we'd need to synchronize the attribute update. On server-side, the server's transport thread can update the attributes before processing any RPCs from the connection.
> 
> So instead of adding Attributes to transportReady(), let's add a new method to ManagedClientTransport.Listener just to do this filtering. It can be called _just before_ transportReady(). With a default implementation that just echos back the same attributes, I hope there's not even many tests that have to be updated.

Thanks for getting back to me! I addressed these suggestions, ptal and let me know if i understood you correctly. As for backporting this to 1.61, I think we're fine with waiting for 1.62 if you folks are at all worried about the risk. Thanks again!

ejona86 commented on binder/src/main/java/io/grpc/binder/internal/BinderTransport.java:
```diff
@@ -758,6 +758,7 @@ private void checkSecurityPolicy(IBinder binder) {
             // triggers), could have shut us down.
             if (!isShutdown()) {
               setState(TransportState.READY);
+              clientTransportListener.filterTransport(attributes);
```
````suggestion
              attributes = clientTransportListener.filterTransport(attributes);
```

ejona86 commented on core/src/test/java/io/grpc/internal/InternalSubchannelTest.java:
```diff
@@ -1360,7 +1361,8 @@ private void createInternalSubchannel(EquivalentAddressGroup ... addrs) {
         logId,
-        new ChannelLoggerImpl(subchannelTracer, fakeClock.getTimeProvider()));
+        new ChannelLoggerImpl(subchannelTracer, fakeClock.getTimeProvider()),
+        Collections.emptyList());
```
```suggestion
          Collections.emptyList());
```

ejona86 commented on core/src/main/java/io/grpc/internal/InternalSubchannel.java:
```diff
@@ -539,6 +544,15 @@ private class TransportListener implements ManagedClientTransport.Listener {
+    public Attributes filterTransport(Attributes attributes) {
+      for (ClientTransportFilter filter : transportFilters) {
+        attributes = Preconditions.checkNotNull(filter.transportReady(attributes),
+        "Filter %s returned null", filter);
```
```suggestion
            "Filter %s returned null", filter);
```

ejona86 commented on api/src/main/java/io/grpc/ClientTransportFilter.java:
```diff
@@ -0,0 +1,51 @@
+   * @return new transport attributes. Default implementation returns the passed-in attributes
+   * intact.
+   */
+  public Attributes transportReady(Attributes transportAttrs){
```
```suggestion
  public Attributes transportReady(Attributes transportAttrs) {
```

ejona86 commented on core/src/main/java/io/grpc/internal/OobChannel.java:
```diff
@@ -130,6 +130,11 @@ public void transportReady() {
         }
 
+        @Override
+        public Attributes filterTransport(Attributes attributes){
```
```suggestion
        public Attributes filterTransport(Attributes attributes) {
```

ejona86 reviewed (approved):
This looks good to me, other than some nits.

sergiitk commented on binder/src/main/java/io/grpc/binder/internal/BinderTransport.java:
````diff
@@ -758,6 +758,8 @@ private void checkSecurityPolicy(IBinder binder) {
             if (!isShutdown()) {
               setState(TransportState.READY);
+              attributes = clientTransportListener.filterTransport(attributes);
+```
````
```suggestion
```

sergiitk commented:
```
/tmpfs/src/github/grpc-java/cronet/src/main/java/io/grpc/cronet/CronetClientTransport.java:172: error: cannot assign a value to final variable attrs
        attrs = CronetClientTransport.this.listener.filterTransport(attrs);
        ^
```

larry-safran commented:
Note to build the android related classes (which is what is catching this),
use `-PskipAndroid=false`

On Tue, Dec 26, 2023 at 3:17 PM Sergii Tkachenko ***@***.***>
wrote:

> /tmpfs/src/github/grpc-java/cronet/src/main/java/io/grpc/cronet/CronetClientTransport.java:172: error: cannot assign a value to final variable attrs
>         attrs = CronetClientTransport.this.listener.filterTransport(attrs);
>         ^
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc-java/pull/10646#issuecomment-1869816363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZQMCXWTIFHPVHQX77ROTVLYLNLHPAVCNFSM6AAAAAA64ZPITOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNRZHAYTMMZWGM>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

larry-safran commented:
@joybestourous The builds failed because of `CronetClientTransportTest`.  That test uses a mock for `clientTransportListener` which by default returns `null` for all of the methods.  You will need to put a `when` clause into `setUp()` so that `clientTransportListener.filterTransport(attr)` just returns its argument.

For example:
```
when(clientTransportListener.filterTransport(any()).thenAnswer(i -> i.getArguments()[0])
```

joybestourous commented:
> @joybestourous The builds failed because of `CronetClientTransportTest`. That test uses a mock for `clientTransportListener` which by default returns `null` for all of the methods. You will need to put a `when` clause into `setUp()` so that `clientTransportListener.filterTransport(attr)` just returns its argument.
> 
> For example:
> 
> ```
> when(clientTransportListener.filterTransport(any()).thenAnswer(i -> i.getArguments()[0])
> ```

Thanks for bringing this to completion 🙏  sorry I missed these errors my android setup was not working properly.

larry-safran commented:
No Problem.  Thanks for creating this and taking care of all of the other
issues.

On Wed, Jan 3, 2024 at 11:20 AM joybestourous ***@***.***>
wrote:

> @joybestourous <https://github.com/joybestourous> The builds failed
> because of CronetClientTransportTest. That test uses a mock for
> clientTransportListener which by default returns null for all of the
> methods. You will need to put a when clause into setUp() so that
> clientTransportListener.filterTransport(attr) just returns its argument.
>
> For example:
>
> when(clientTransportListener.filterTransport(any()).thenAnswer(i -> i.getArguments()[0])
>
> Thanks for bringing this to completion 🙏 sorry I missed these errors my
> android setup was not working properly.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc-java/pull/10646#issuecomment-1875850695>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZQMCXR3L3PK45TYNWWB66TYMWVQRAVCNFSM6AAAAAA64ZPITOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNZVHA2TANRZGU>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

ejona86 commented:
#10805 is suspicious, and shouldn't be necessary. I feel like there's no harm happening at present with making the field volatile, but the fact TSAN is finding issues makes clear something is happening that we don't expect. So we aren't going to backport this into the 1.61 release. We'll probably try to remove the volatile before 1.62, but even if we don't, at that point it will have seen enough testing that we are then more confident it won't cause an issue.

joybestourous commented:
> #10805 is suspicious, and shouldn't be necessary. I feel like there's no harm happening at present with making the field volatile, but the fact TSAN is finding issues makes clear something is happening that we don't expect. So we aren't going to backport this into the 1.61 release. We'll probably try to remove the volatile before 1.62, but even if we don't, at that point it will have seen enough testing that we are then more confident it won't cause an issue.

Oh hm that's super surprising / suspicious. Makes complete sense to put this in 1.62. Thanks for the heads up!
